### PR TITLE
Weak Equivalence between Categories

### PR DESCRIPTION
**Pull request by kangrongji**
This PR adds the notion of weak equivalence of category, i.e. functors that are both fully-faithful and essentially surjective. The main result is that, weak equivalence between univalent categories is equivalence, namely it admits explicit inverse functor.

P.S. I change the definition of isomorphisms in category `CatIso`, by separating it into a record predicate `isIso` and re-define `CatIso` as a Σ-type. This allows some definitions to be more directly formulated, and some arguments easier. Forgive me if this makes review more painful...

P.P.S. Instead of the brutal-force argument in HoTT book, I use a dependent version of isomorphism to formulate a more structural proof. Price is a new file `Cubical.Foundation.Equiv.Dependent`.

### CODE REVIEW

File: Cubical/Categories/Constructions/BinProduct.agda [hunk 1]
```diff
@@ -62,3 +63,14 @@ module _ (C : Category ℓC ℓC')
     - define inverse to `assoc`, prove isomorphism
     - prove product is commutative up to isomorphism
 -}
+
+
+  -- The isomorphisms in product category
+
+  open isIso
+
+  iso× : {x y : C .ob}{z w : D .ob} → CatIso C x y → CatIso D z w → CatIso (C × D) (x , z) (y , w)
```
Comment (mortberg) [inline Cubical/Categories/Constructions/BinProduct.agda:72]: Please rename to `CatIso×`

File: Cubical/Categories/Category/Base.agda [hunk 2]
```diff
@@ -51,40 +52,63 @@ comp' = _∘_
 infixr 16 comp'
 syntax comp' C g f = g ∘⟨ C ⟩ f
 
+
 -- Isomorphisms and paths in categories
-record CatIso (C : Category ℓ ℓ') (x y : C .ob) : Type ℓ' where
-  constructor catiso
+
+record isIso (C : Category ℓ ℓ'){x y : C .ob}(f : C [ x , y ]) : Type ℓ' where
```
Comment (mortberg) [inline Cubical/Categories/Category/Base.agda:58]: Maybe we should have no-eta-equality for this? Might be good to avoid slow typechecking times

**Review by mortberg — CHANGES_REQUESTED** [review]
Two small comments for things to consider changing. Apart from this it looks great! Sorry for the slow review. I'll take a look at your later PRs once this is merged.

This seems like a good refactoring of CatIso and some useful stuff about dependent isos/equivs

File: Cubical/Categories/Category/Base.agda (hunk 2 above)
Comment (kangrongji) [reply Cubical/Categories/Category/Base.agda:58]: Oh, the extra judgemental equality is very useful, e.g. [symmetry of inverse](https://github.com/agda/cubical/blob/cc8812edc0d3ac57e60cfebe5dd42c49fcb773ff/Cubical/Categories/Morphism.agda#L129), and I think it is used in many places in the library.

Comment (kangrongji) [reply Cubical/Categories/Category/Base.agda:58]: The typechecking time is not a problem so far, do you think it is necessary to change it?

Comment (mortberg) [reply Cubical/Categories/Category/Base.agda:58]: Ah ok, if it's used several times and simplifies proofs but doesn't lead to any problems for typechecking times then we should not remove it

**Review by mortberg — APPROVED** [review]

**Pull request merged**